pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,984 @@
+	// will be advertised regardless.
+	if noTrafficPolicyLocal(svc) {
+		return
+	}
+	if c.hasAffectedBPsByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice ADD event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateEndpointSlice(_, obj interface{}) {
+	eps := obj.(*discovery.EndpointSlice)
+	svc, _ := c.serviceLister.Services(eps.GetNamespace()).Get(eps.GetLabels()[discovery.LabelServiceName])
+	if svc == nil {
+		return
+	}
+	// Events for EndpointSlices dedicated to Services without a `Local` traffic policy are ignored, as the Service IPs
+	// will be advertised regardless.
+	if noTrafficPolicyLocal(svc) {
+		return
+	}
+	if c.hasAffectedBPsByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice UPDATE event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) hasAffectedBPsByEgress() bool {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if !c.matchedCurrentNode(bp) {
+			continue
+		}
+		if bp.Spec.Advertisements.Egress != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) addEgress(obj interface{}) {
+	eg := obj.(*v1beta1.Egress)
+	if eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPsByEgress() {
+		klog.V(2).InfoS("Processing Egress ADD event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateEgress(oldObj, obj interface{}) {
+	oldEg := oldObj.(*v1beta1.Egress)
+	eg := obj.(*v1beta1.Egress)
+	if oldEg.Status.EgressNode != c.nodeName && eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if oldEg.Status.EgressIP == eg.Status.EgressIP {
+		return
+	}
+	if c.hasAffectedBPsByEgress() {
+		klog.V(2).InfoS("Processing Egress UPDATE event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) deleteEgress(obj interface{}) {
+	eg := obj.(*v1beta1.Egress)
+	if eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPsByEgress() {
+		klog.V(2).InfoS("Processing Egress DELETE event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) filterAffectedBPsByNode(node *corev1.Node) sets.Set[string] {
+	affectedBPs := sets.New[string]()
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if matchedNode(node, bp) {
+			affectedBPs.Insert(bp.GetName())
+		}
+	}
+	return affectedBPs
+}
+
+func (c *Controller) updateNode(oldObj, obj interface{}) {
+	oldNode := oldObj.(*corev1.Node)
+	node := obj.(*corev1.Node)
+	if node.GetName() != c.nodeName {
+		return
+	}
+	if reflect.DeepEqual(node.GetLabels(), oldNode.GetLabels()) &&
+		reflect.DeepEqual(node.GetAnnotations(), oldNode.GetAnnotations()) {
+		return
+	}
+	oldAffectedBPs := c.filterAffectedBPsByNode(oldNode)
+	newAffectedBPs := c.filterAffectedBPsByNode(node)
+	if oldAffectedBPs.Len() != 0 || newAffectedBPs.Len() != 0 {
```

Comment:
ditto. why comparing number of affected policies?

Reply:
Updated to use `hasAffectedBPByNode`